File: android/app/src/main/java/com/gymapp/tracker/core/domain/Calculations.kt
package com.gymapp.tracker.core.domain

import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Training mathematics — the Kotlin port of `backend/src/domain/calculations.ts`.
 *
 * The formulas must stay identical to the backend's: if the two ever disagree,
 * the same workout shows different numbers depending on where it was computed.
 * `docs/CALCULATIONS.md` documents the reasoning behind each one.
 */

/** Minimal view of a set — everything the maths needs, nothing else. */
data class SetValues(
    val weightKg: Double? = null,
    val reps: Int? = null,
    val durationSec: Int? = null,
    val distanceM: Double? = null,
    val isWarmup: Boolean = false,
    val isOneRmTest: Boolean = false,
)

data class SetSummary(
    val sets: Int = 0,
    val reps: Int = 0,
    val volumeKg: Double = 0.0,
    val maxWeightKg: Double? = null,
    val durationSec: Int = 0,
    val distanceM: Double = 0.0,
)

data class TrendResult(
    val changePercent: Double? = null,
    /** False when there was too little data to say anything meaningful. */
    val reliable: Boolean = false,
    val previousMedian: Double? = null,
    val currentMedian: Double? = null,
)

data class BestE1rm(val value: Double, val set: SetValues?)

/** Volume of a single set. Only counts when both weight and reps are present. */
fun setVolume(set: SetValues): Double {
    val weight = set.weightKg ?: return 0.0
    val reps = set.reps ?: return 0.0
    if (weight <= 0 || reps <= 0) return 0.0
    return weight * reps
}

/** Aggregates a list of sets. Warm-up sets can be excluded from the totals. */
fun summarizeSets(sets: List<SetValues>, includeWarmups: Boolean = true): SetSummary {
    val relevant = if (includeWarmups) sets else sets.filterNot { it.isWarmup }
    var volume = 0.0
    var reps = 0
    var duration = 0
    var distance = 0.0
    var maxWeight: Double? = null

    for (set in relevant) {
        volume += setVolume(set)
        reps += set.reps?.takeIf { it > 0 } ?: 0
        duration += set.durationSec?.takeIf { it > 0 } ?: 0
        distance += set.distanceM?.takeIf { it > 0 } ?: 0.0
        val weight = set.weightKg
        if (weight != null && weight > 0 && (maxWeight == null || weight > maxWeight!!)) maxWeight = weight
    }

    return SetSummary(
        sets = relevant.size,
        reps = reps,
        volumeKg = round2(volume),
        maxWeightKg = maxWeight,
        durationSec = duration,
        distanceM = round2(distance),
    )
}

/** Reps above this are capped — Epley overestimates badly in high rep ranges. */
const val E1RM_REP_CAP = 12

/**
 * Estimated one-rep max, Epley: weight × (1 + reps / 30).
 * Returns 0 for input that cannot produce a meaningful estimate.
 */
fun estimateOneRepMax(weightKg: Double?, reps: Int?): Double {
    if (weightKg == null || reps == null) return 0.0
    if (weightKg <= 0 || reps <= 0) return 0.0
    if (reps == 1) return round2(weightKg)
    val capped = minOf(reps, E1RM_REP_CAP)
    return round2(weightKg * (1 + capped / 30.0))
}

/** True when the rep count was above the cap, so the estimate is extrapolated. */
fun isE1rmExtrapolated(reps: Int?): Boolean = (reps ?: 0) > E1RM_REP_CAP

/** Best estimated 1RM across a list of sets, with the set that produced it. */
fun bestE1rm(sets: List<SetValues>): BestE1rm {
    var best = 0.0
    var bestSet: SetValues? = null
    for (set in sets) {
        if (set.isWarmup) continue
        val value = estimateOneRepMax(set.weightKg, set.reps)
        if (value > best) {
            best = value
            bestSet = set
        }
    }
    return BestE1rm(best, bestSet)
}

/**
 * One comparable performance number per set, so a exercise can be compared with
 * itself over time regardless of how it is measured.
 */
fun performanceMetric(set: SetValues): Double? = when {
    set.weightKg != null && set.weightKg > 0 && (set.reps ?: 0) > 0 -> estimateOneRepMax(set.weightKg, set.reps)
    (set.reps ?: 0) > 0 -> set.reps!!.toDouble()
    (set.distanceM ?: 0.0) > 0 && (set.durationSec ?: 0) > 0 -> set.distanceM!! / set.durationSec!!
    (set.distanceM ?: 0.0) > 0 -> set.distanceM
    (set.durationSec ?: 0) > 0 -> set.durationSec!!.toDouble()
    else -> null
}

// --- statistics helpers ----------------------------------------------------

fun mean(values: List<Double>): Double =
    if (values.isEmpty()) 0.0 else round2(values.sum() / values.size)

fun median(values: List<Double>): Double {
    if (values.isEmpty()) return 0.0
    val sorted = values.sorted()
    val middle = sorted.size / 2
    return round2(
        if (sorted.size % 2 == 0) (sorted[middle - 1] + sorted[middle]) / 2 else sorted[middle],
    )
}

/** Simple trailing moving average; the first entries average what exists so far. */
fun movingAverage(values: List<Double>, window: Int): List<Double> {
    if (window <= 1) return values
    return values.indices.map { index ->
        val from = maxOf(0, index - window + 1)
        round2(values.subList(from, index + 1).average())
    }
}

/** Percentage change, or null when there is no sensible denominator. */
fun percentChange(previous: Double?, current: Double?): Double? {
    if (previous == null || current == null) return null
    if (previous == 0.0) return null
    return round1(((current - previous) / abs(previous)) * 100)
}

/**
 * Trend between two periods, resistant to single outliers.
 *
 * Uses the median rather than the mean and requires at least two data points on
 * each side; otherwise one record set would read as lasting progress.
 */
fun robustTrend(previous: List<Double>, current: List<Double>, minSamples: Int = 2): TrendResult {
    if (previous.size < minSamples || current.size < minSamples) {
        return TrendResult(changePercent = null, reliable = false)
    }
    val previousMedian = median(previous)
    val currentMedian = median(current)
    return TrendResult(
        changePercent = percentChange(previousMedian, currentMedian),
        reliable = true,
        previousMedian = previousMedian,
        currentMedian = currentMedian,
    )
}

/** Consecutive sets with identical values, for the "100 kg × 10 × 3" summary. */
data class SetGroup(val set: SetValues, val count: Int)

fun groupSets(sets: List<SetValues>): List<SetGroup> {
    val groups = mutableListOf<SetGroup>()
    for (set in sets) {
        val last = groups.lastOrNull()
        if (last != null && sameShape(last.set, set)) {
            groups[groups.lastIndex] = last.copy(count = last.count + 1)
        } else {
            groups.add(SetGroup(set, 1))
        }
    }
    return groups
}

private fun sameShape(a: SetValues, b: SetValues): Boolean =
    a.weightKg == b.weightKg &&
        a.reps == b.reps &&
        a.durationSec == b.durationSec &&
        a.distanceM == b.distanceM

internal fun round1(value: Double): Double = (value * 10).roundToInt() / 10.0
internal fun round2(value: Double): Double = (value * 100).roundToInt() / 100.0

File: backend/src/domain/calculations.ts
import { ExerciseType } from '@prisma/client';
import { round } from '../lib/units';

/**
 * Pure, dependency-free workout mathematics. Everything in here is unit tested
 * (tests/calculations.test.ts) and documented in docs/CALCULATIONS.md.
 */

export interface SetLike {
  weightKg?: number | null;
  reps?: number | null;
  durationSec?: number | null;
  distanceM?: number | null;
  isWarmup?: boolean;
  isOneRmTest?: boolean;
}

// ---------------------------------------------------------------------------
// Volume
// ---------------------------------------------------------------------------

/**
 * Volume load of a single set: `weight × reps`.
 *
 * Only sets that have both a weight and a rep count contribute to kilogram
 * volume. Body-weight exercises without additional load have no meaningful
 * kg-volume – they are counted through `totalReps` instead, so that a set of
 * 20 push-ups never silently inflates the tonnage of a bench press session.
 */
export function setVolume(set: SetLike): number {
  if (set.weightKg == null || set.reps == null) return 0;
  if (set.weightKg <= 0 || set.reps <= 0) return 0;
  return set.weightKg * set.reps;
}

export interface VolumeSummary {
  volumeKg: number;
  sets: number;
  reps: number;
  workingSets: number;
  durationSec: number;
  distanceM: number;
}

export function summarizeSets(sets: SetLike[], includeWarmups = true): VolumeSummary {
  const relevant = includeWarmups ? sets : sets.filter((s) => !s.isWarmup);
  return relevant.reduce<VolumeSummary>(
    (acc, set) => {
      acc.volumeKg += setVolume(set);
      acc.sets += 1;
      acc.reps += set.reps ?? 0;
      acc.workingSets += set.isWarmup ? 0 : 1;
      acc.durationSec += set.durationSec ?? 0;
      acc.distanceM += set.distanceM ?? 0;
      return acc;
    },
    { volumeKg: 0, sets: 0, reps: 0, workingSets: 0, durationSec: 0, distanceM: 0 },
  );
}

// ---------------------------------------------------------------------------
// One rep max
// ---------------------------------------------------------------------------

/**
 * Estimated one rep maximum using the **Epley formula**:
 *
 *     e1RM = weight × (1 + reps / 30)
 *
 * A single rep returns the weight itself. The estimate loses accuracy above
 * roughly 12 repetitions, therefore rep counts are clamped to 12 for the
 * estimate – this is deliberately conservative and documented in the app.
 */
export const E1RM_REP_CAP = 12;

export function estimateOneRepMax(weightKg: number, reps: number): number {
  if (!Number.isFinite(weightKg) || !Number.isFinite(reps)) return 0;
  if (weightKg <= 0 || reps <= 0) return 0;
  if (reps === 1) return round(weightKg, 2);
  const effectiveReps = Math.min(reps, E1RM_REP_CAP);
  return round(weightKg * (1 + effectiveReps / 30), 2);
}

/** True when the estimate had to clamp the rep count (shown as a hint in the UI). */
export function isE1rmExtrapolated(reps: number): boolean {
  return reps > E1RM_REP_CAP;
}

export function bestE1rm(sets: SetLike[]): { value: number; set: SetLike | null } {
  let best = 0;
  let bestSet: SetLike | null = null;
  for (const set of sets) {
    if (set.weightKg == null || set.reps == null) continue;
    const value = estimateOneRepMax(set.weightKg, set.reps);
    if (value > best) {
      best = value;
      bestSet = set;
    }
  }
  return { value: round(best, 2), set: bestSet };
}

// ---------------------------------------------------------------------------
// Statistics helpers
// ---------------------------------------------------------------------------

export function mean(values: number[]): number {
  if (values.length === 0) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

export function median(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid];
}

/** Simple trailing moving average; window is clamped to the available data. */
export function movingAverage(values: number[], window: number): number[] {
  if (window <= 1) return [...values];
  return values.map((_, i) => {
    const from = Math.max(0, i - window + 1);
    return mean(values.slice(from, i + 1));
  });
}

/**
 * Percentage change between two values. Returns `null` when there is no
 * meaningful baseline (0 or missing previous value) instead of `Infinity`,
 * so the UI can render "–" rather than a nonsense "+∞ %".
 */
export function percentChange(previous: number | null | undefined, current: number | null | undefined): number | null {
  if (previous == null || current == null) return null;
  if (previous === 0) return null;
  return round(((current - previous) / Math.abs(previous)) * 100, 1);
}

/**
 * Robust trend between two sample windows.
 *
 * Raw averages are extremely sensitive to a single outlier session, so the
 * comparison uses the **median** of each window and additionally requires a
 * minimum sample size. Below that size the trend is reported as `null`
 * ("not enough data") rather than a confident looking number.
 */
export interface TrendResult {
  changePercent: number | null;
  previous: number | null;
  current: number | null;
  sampleSize: number;
  reliable: boolean;
}

export const MIN_TREND_SAMPLES = 2;

export function robustTrend(previousWindow: number[], currentWindow: number[]): TrendResult {
  const previous = previousWindow.length ? round(median(previousWindow), 2) : null;
  const current = currentWindow.length ? round(median(currentWindow), 2) : null;
  const sampleSize = Math.min(previousWindow.length, currentWindow.length);
  const reliable = previousWindow.length >= MIN_TREND_SAMPLES && currentWindow.length >= MIN_TREND_SAMPLES;
  return {
    changePercent: reliable ? percentChange(previous, current) : null,
    previous,
    current,
    sampleSize,
    reliable,
  };
}

// ---------------------------------------------------------------------------
// Exercise type helpers
// ---------------------------------------------------------------------------

export function isWeightBased(type: ExerciseType): boolean {
  return type === ExerciseType.STRENGTH || type === ExerciseType.BODYWEIGHT;
}

export function supportsOneRepMax(type: ExerciseType): boolean {
  return type === ExerciseType.STRENGTH;
}

/** Formats "100 kg × 10 × 3" style summaries used by bot + app. */
export function formatSetGroup(weightKg: number | null | undefined, reps: number | null | undefined, count: number): string {
  const parts: string[] = [];
  if (weightKg != null && weightKg > 0) parts.push(`${round(weightKg, 2)} kg`);
  if (reps != null && reps > 0) parts.push(`${reps}`);
  if (count > 1) parts.push(`${count}`);
  return parts.join(' × ');
}

/**
 * Groups consecutive identical sets so that "100/10, 100/10, 100/10" is shown
 * as "100 kg × 10 × 3" while "100/10, 110/8" stays expanded.
 */
export interface SetGroup {
  weightKg: number | null;
  reps: number | null;
  durationSec: number | null;
  distanceM: number | null;
  count: number;
}

export function groupSets(sets: SetLike[]): SetGroup[] {
  const groups: SetGroup[] = [];
  for (const set of sets) {
    const last = groups[groups.length - 1];
    if (
      last &&
      last.weightKg === (set.weightKg ?? null) &&
      last.reps === (set.reps ?? null) &&
      last.durationSec === (set.durationSec ?? null) &&
      last.distanceM === (set.distanceM ?? null)
    ) {
      last.count += 1;
    } else {
      groups.push({
        weightKg: set.weightKg ?? null,
        reps: set.reps ?? null,
        durationSec: set.durationSec ?? null,
        distanceM: set.distanceM ?? null,
        count: 1,
      });
    }
  }
  return groups;
}
